importer - record/field mismatch

clang crash while analyzing VIM vim/src/buffer.c

analyze: INFO: execute compiler: gcc -c -I/ssd/ednikru/ctu/tests/vim/src -I/ssd/ednikru/ctu/tests/vim/src/proto -DHAVE_CONFIG_H -DFEAT_GUI_GTK -pthread -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/pango-1.0 -I/usr/include/gio-unix-2.0 -I/usr/include/freetype2 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/libpng12 -I/usr/include/harfbuzz -I/usr/include/c++/4.8 -I/usr/include/x86_64-linux-gnu/c++/4.8 -I/usr/include/c++/4.8/backward -I/usr/local/include -I/usr/include/x86_64-linux-gnu -I/usr/include -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -o /ssd/ednikru/ctu/tests/vim/src/objects/buffer.o /ssd/ednikru/ctu/tests/vim/src/buffer.c
analyze: INFO: Importing function _Z10prep_exargP5exargP11file_buffer@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/fileio.c.ast
analyze: INFO: Importing function _Z5allocj@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/misc2.c.ast
analyze: INFO: Importing function _Z6lallocmi@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/misc2.c.ast
analyze: INFO: Importing function _Z13clear_sb_textv@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/message.c.ast
analyze: INFO: Importing function _Z14mf_release_allv@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/memfile.c.ast
analyze: INFO: Importing function _Z15do_outofmem_msgm@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/misc2.c.ast
analyze: INFO: Importing function _Z14block_autocmdsv@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/fileio.c.ast
analyze: INFO: Importing function _Z16unblock_autocmdsv@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/fileio.c.ast
analyze: INFO: Importing function _Z15get_vim_var_stri@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/eval.c.ast
analyze: INFO: Importing function _Z13get_tv_stringP8typval_T@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/eval.c.ast
analyze: INFO: Importing function _Z17get_tv_string_bufP8typval_TPh@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/eval.c.ast
analyze: INFO: Importing function _Z21get_tv_string_buf_chkP8typval_TPh@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/eval.c.ast
analyze: INFO: Importing function _Z5emsgnPhl@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/message.c.ast
analyze: INFO: Importing function _Z12vim_snprintfPcmS_z@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/message.c.ast
analyze: INFO: Importing function _Z13aucmd_prepbufP10aco_save_TP11file_buffer@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/fileio.c.ast
analyze: INFO: Importing function _Z14win_init_emptyP8window_S@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/window.c.ast
analyze: INFO: Importing function _Z16redraw_win_laterP8window_Si@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/screen.c.ast
analyze: INFO: Importing function _Z8vim_freePv@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/misc2.c.ast
analyze: INFO: Importing function _Z13make_snapshoti@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/window.c.ast
analyze: INFO: Importing function _Z11alloc_clearj@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/misc2.c.ast
analyze: INFO: Importing function _Z13win_split_insiiP8window_Si@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/window.c.ast
analyze: INFO: Importing function _Z12win_comp_posv@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/window.c.ast
analyze: INFO: Importing function _Z14tabline_heightv@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/window.c.ast
analyze: INFO: Importing function _Z15gui_use_tablinev@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/gui.c.ast
analyze: INFO: Importing function _Z7ml_openP11file_buffer@x86_64 from /ssd/ednikru/ctu/tests/vim/.xtu//ast/x86_64/ssd/ednikru/ctu/tests/vim/src/memline.c.ast
analyze: INFO: clang-4.0: /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:4798: bool {anonymous}::LValueExprEvaluatorBase<Derived>::VisitMemberExpr(const clang::MemberExpr*) [with Derived = {anonymous}::LValueExprEvaluator]: Assertion `BaseTy->getAs<RecordType>()->getDecl()->getCanonicalDecl() == FD->getParent()->getCanonicalDecl() && "record / field mismatch"' failed.
analyze: INFO: #0 0x00000000037e0f7d llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/ednikru/ssd/ctu/llvm/lib/Support/Unix/Signals.inc:402:0
analyze: INFO: #1 0x00000000037e12d8 PrintStackTraceSignalHandler(void*) /home/ednikru/ssd/ctu/llvm/lib/Support/Unix/Signals.inc:466:0
analyze: INFO: #2 0x00000000037df49e llvm::sys::RunSignalHandlers() /home/ednikru/ssd/ctu/llvm/lib/Support/Signals.cpp:44:0
analyze: INFO: #3 0x00000000037e090f SignalHandler(int) /home/ednikru/ssd/ctu/llvm/lib/Support/Unix/Signals.inc:256:0
analyze: INFO: #4 0x00007f64e2597330 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x10330)
analyze: INFO: #5 0x00007f64e1392c37 gsignal /build/eglibc-oGUzwX/eglibc-2.19/signal/../nptl/sysdeps/unix/sysv/linux/raise.c:56:0
analyze: INFO: #6 0x00007f64e1396028 abort /build/eglibc-oGUzwX/eglibc-2.19/stdlib/abort.c:91:0
analyze: INFO: #7 0x00007f64e138bbf6 __assert_fail_base /build/eglibc-oGUzwX/eglibc-2.19/assert/assert.c:92:0
analyze: INFO: #8 0x00007f64e138bca2 (/lib/x86_64-linux-gnu/libc.so.6+0x2fca2)
analyze: INFO: #9 0x000000000645403b (anonymous namespace)::LValueExprEvaluatorBase<(anonymous namespace)::LValueExprEvaluator>::VisitMemberExpr(clang::MemberExpr const*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:4800:0
analyze: INFO: #10 0x0000000006434385 (anonymous namespace)::LValueExprEvaluator::VisitMemberExpr(clang::MemberExpr const*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:5094:0
analyze: INFO: #11 0x000000000645325b clang::StmtVisitorBase<clang::make_const_ptr, (anonymous namespace)::LValueExprEvaluator, bool>::Visit(clang::Stmt const*) /home/ednikru/ssd/ctu/build_ctu_master/tools/clang/include/clang/AST/StmtNodes.inc:591:0
analyze: INFO: #12 0x000000000643331b EvaluateLValue(clang::Expr const*, (anonymous namespace)::LValue&, (anonymous namespace)::EvalInfo&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:4941:0
analyze: INFO: #13 0x00000000064610a3 (anonymous namespace)::ExprEvaluatorBase<(anonymous namespace)::IntExprEvaluator>::VisitCastExpr(clang::CastExpr const*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:4655:0
analyze: INFO: #14 0x0000000006444905 (anonymous namespace)::IntExprEvaluator::VisitCastExpr(clang::CastExpr const*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:8622:0
analyze: INFO: #15 0x000000000647d5db clang::StmtVisitorBase<clang::make_const_ptr, (anonymous namespace)::IntExprEvaluator, bool>::VisitImplicitCastExpr(clang::ImplicitCastExpr const*) /home/ednikru/ssd/ctu/build_ctu_master/tools/clang/include/clang/AST/StmtNodes.inc:423:0
analyze: INFO: #16 0x000000000645fabb clang::StmtVisitorBase<clang::make_const_ptr, (anonymous namespace)::IntExprEvaluator, bool>::Visit(clang::Stmt const*) /home/ednikru/ssd/ctu/build_ctu_master/tools/clang/include/clang/AST/StmtNodes.inc:423:0
analyze: INFO: #17 0x0000000006449e60 Evaluate(clang::APValue&, (anonymous namespace)::EvalInfo&, clang::Expr const*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:9567:0
analyze: INFO: #18 0x000000000644a790 EvaluateAsRValue((anonymous namespace)::EvalInfo&, clang::Expr const*, clang::APValue&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:9656:0
analyze: INFO: #19 0x000000000644abab clang::Expr::EvaluateAsRValue(clang::Expr::EvalResult&, clang::ASTContext const&) const /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:9711:0
analyze: INFO: #20 0x000000000644ac2c clang::Expr::EvaluateAsBooleanCondition(bool&, clang::ASTContext const&) const /home/ednikru/ssd/ctu/llvm/tools/clang/lib/AST/ExprConstant.cpp:9717:0
analyze: INFO: #21 0x00000000061bee0e (anonymous namespace)::CFGBuilder::evaluateAsBooleanConditionNoCache(clang::Expr*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Analysis/CFG.cpp:955:0
analyze: INFO: #22 0x00000000061beaae (anonymous namespace)::CFGBuilder::tryEvaluateBool(clang::Expr*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Analysis/CFG.cpp:907:0
analyze: INFO: #23 0x00000000061c315e (anonymous namespace)::CFGBuilder::VisitIfStmt(clang::IfStmt*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Analysis/CFG.cpp:2272:0
analyze: INFO: #24 0x00000000061c0c7c (anonymous namespace)::CFGBuilder::Visit(clang::Stmt*, (anonymous namespace)::AddStmtChoice) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Analysis/CFG.cpp:1544:0
analyze: INFO: #25 0x00000000061bd427 (anonymous namespace)::CFGBuilder::addStmt(clang::Stmt*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Analysis/CFG.cpp:578:0
analyze: INFO: #26 0x00000000061c22de (anonymous namespace)::CFGBuilder::VisitCompoundStmt(clang::CompoundStmt*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Analysis/CFG.cpp:1965:0
analyze: INFO: #27 0x00000000061c09b0 (anonymous namespace)::CFGBuilder::Visit(clang::Stmt*, (anonymous namespace)::AddStmtChoice) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Analysis/CFG.cpp:1474:0
analyze: INFO: #28 0x00000000061bd427 (anonymous namespace)::CFGBuilder::addStmt(clang::Stmt*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Analysis/CFG.cpp:578:0
analyze: INFO: #29 0x00000000061bf20e (anonymous namespace)::CFGBuilder::buildCFG(clang::Decl const*, clang::Stmt*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Analysis/CFG.cpp:1039:0
analyze: INFO: #30 0x00000000061c8494 clang::CFG::buildCFG(clang::Decl const*, clang::Stmt*, clang::ASTContext*, clang::CFG::BuildOptions const&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Analysis/CFG.cpp:3897:0
analyze: INFO: #31 0x00000000061ad6bc clang::AnalysisDeclContext::getCFG() /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Analysis/AnalysisDeclContext.cpp:217:0
analyze: INFO: #32 0x00000000055939eb mayInlineDecl(clang::AnalysisDeclContext*, clang::AnalyzerOptions&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngineCallAndReturn.cpp:773:0
analyze: INFO: #33 0x0000000005593c19 clang::ento::ExprEngine::shouldInlineCall(clang::ento::CallEvent const&, clang::Decl const*, clang::ento::ExplodedNode const*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngineCallAndReturn.cpp:825:0
analyze: INFO: #34 0x0000000005594110 clang::ento::ExprEngine::defaultEvalCall(clang::ento::NodeBuilder&, clang::ento::ExplodedNode*, clang::ento::CallEvent const&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngineCallAndReturn.cpp:914:0
analyze: INFO: #35 0x0000000005539a1b clang::ento::CheckerManager::runCheckersForEvalCall(clang::ento::ExplodedNodeSet&, clang::ento::ExplodedNodeSet const&, clang::ento::CallEvent const&, clang::ento::ExprEngine&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/CheckerManager.cpp:617:0
analyze: INFO: #36 0x0000000005592c6e clang::ento::ExprEngine::evalCall(clang::ento::ExplodedNodeSet&, clang::ento::ExplodedNode*, clang::ento::CallEvent const&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngineCallAndReturn.cpp:519:0
analyze: INFO: #37 0x0000000005592b0b clang::ento::ExprEngine::VisitCallExpr(clang::CallExpr const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngineCallAndReturn.cpp:485:0
analyze: INFO: #38 0x0000000005567e5e clang::ento::ExprEngine::Visit(clang::Stmt const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:1179:0
analyze: INFO: #39 0x0000000005564aa0 clang::ento::ExprEngine::ProcessStmt(clang::CFGStmt, clang::ento::ExplodedNode*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:507:0
analyze: INFO: #40 0x0000000005563da0 clang::ento::ExprEngine::processCFGElement(clang::CFGElement, clang::ento::ExplodedNode*, unsigned int, clang::ento::NodeBuilderContext*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:355:0
analyze: INFO: #41 0x0000000005552b2d clang::ento::CoreEngine::HandlePostStmt(clang::CFGBlock const*, unsigned int, clang::ento::ExplodedNode*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:532:0
analyze: INFO: #42 0x00000000055519a0 clang::ento::CoreEngine::dispatchWorkItem(clang::ento::ExplodedNode*, clang::ProgramPoint, clang::ento::WorkListUnit const&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:279:0
analyze: INFO: #43 0x0000000005551580 clang::ento::CoreEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int, llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:216:0
analyze: INFO: #44 0x0000000005034dea clang::ento::ExprEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int) /home/ednikru/ssd/ctu/llvm/tools/clang/include/clang/StaticAnalyzer/Core/PathSensitive/ExprEngine.h:110:0
analyze: INFO: #45 0x0000000004feccc4 (anonymous namespace)::AnalysisConsumer::ActionExprEngine(clang::Decl*, bool, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*> >*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:826:0
analyze: INFO: #46 0x0000000004fed0b4 (anonymous namespace)::AnalysisConsumer::RunPathSensitiveChecks(clang::Decl*, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*> >*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:862:0
analyze: INFO: #47 0x0000000004feca8a (anonymous namespace)::AnalysisConsumer::HandleCode(clang::Decl*, unsigned int, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*> >*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:792:0
analyze: INFO: #48 0x0000000004feb401 (anonymous namespace)::AnalysisConsumer::HandleDeclsCallGraph(unsigned int) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:575:0
analyze: INFO: #49 0x0000000004febe49 (anonymous namespace)::AnalysisConsumer::HandleTranslationUnit(clang::ASTContext&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:657:0
analyze: INFO: #50 0x0000000005628933 clang::ParseAST(clang::Sema&, bool, bool) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Parse/ParseAST.cpp:161:0
analyze: INFO: #51 0x000000000401610a clang::ASTFrontendAction::ExecuteAction() /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Frontend/FrontendAction.cpp:560:0
analyze: INFO: #52 0x0000000004015baf clang::FrontendAction::Execute() /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Frontend/FrontendAction.cpp:463:0
analyze: INFO: #53 0x0000000003fbcd73 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/Frontend/CompilerInstance.cpp:954:0
analyze: INFO: #54 0x000000000413d9ee clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/ednikru/ssd/ctu/llvm/tools/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:249:0
analyze: INFO: #55 0x0000000001cb64c1 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/ednikru/ssd/ctu/llvm/tools/clang/tools/driver/cc1_main.cpp:221:0
analyze: INFO: #56 0x0000000001cac87a ExecuteCC1Tool(llvm::ArrayRef<char const*>, llvm::StringRef) /home/ednikru/ssd/ctu/llvm/tools/clang/tools/driver/driver.cpp:299:0
analyze: INFO: #57 0x0000000001cad478 main /home/ednikru/ssd/ctu/llvm/tools/clang/tools/driver/driver.cpp:380:0
analyze: INFO: #58 0x00007f64e137df45 __libc_start_main /build/eglibc-oGUzwX/eglibc-2.19/csu/libc-start.c:321:0
analyze: INFO: #59 0x0000000001caa2d9 _start (/ssd/ednikru/ctu/build_ctu_master/bin/clang-4.0+0x1caa2d9)
analyze: INFO: Stack dump:
analyze: INFO: 0.	Program arguments: /ssd/ednikru/ctu/build_ctu_master/bin/clang-4.0 -cc1 -triple x86_64-unknown-linux-gnu -analyze -disable-free -main-file-name buffer.c -analyzer-store=region -analyzer-opt-analyze-nested-blocks -analyzer-eagerly-assume -analyzer-checker=core -analyzer-checker=apiModeling -analyzer-checker=unix -analyzer-checker=deadcode -analyzer-checker=security.insecureAPI.UncheckedReturn -analyzer-checker=security.insecureAPI.getpw -analyzer-checker=security.insecureAPI.gets -analyzer-checker=security.insecureAPI.mktemp -analyzer-checker=security.insecureAPI.mkstemp -analyzer-checker=security.insecureAPI.vfork -analyzer-checker=nullability.NullPassedToNonnull -analyzer-checker=nullability.NullReturnedFromNonnull -analyzer-output plist-multi-file -w -analyzer-config xtu-dir=/ssd/ednikru/ctu/tests/vim/.xtu -analyzer-config reanalyze-xtu-visited=true -analyzer-stats -mrelocation-model static -mthread-model posix -fmath-errno -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu x86-64 -momit-leaf-frame-pointer -dwarf-column-info -debugger-tuning=gdb -resource-dir /ssd/ednikru/ctu/build_ctu_master/bin/../lib/clang/4.0.0 -I /ssd/ednikru/ctu/tests/vim/src -I /ssd/ednikru/ctu/tests/vim/src/proto -D HAVE_CONFIG_H -D FEAT_GUI_GTK -I /usr/include/gtk-2.0 -I /usr/lib/x86_64-linux-gnu/gtk-2.0/include -I /usr/include/atk-1.0 -I /usr/include/cairo -I /usr/include/gdk-pixbuf-2.0 -I /usr/include/pango-1.0 -I /usr/include/gio-unix-2.0 -I /usr/include/freetype2 -I /usr/include/glib-2.0 -I /usr/lib/x86_64-linux-gnu/glib-2.0/include -I /usr/include/pixman-1 -I /usr/include/libpng12 -I /usr/include/harfbuzz -I /usr/include/c++/4.8 -I /usr/include/x86_64-linux-gnu/c++/4.8 -I /usr/include/c++/4.8/backward -I /usr/local/include -I /usr/include/x86_64-linux-gnu -I /usr/include -U _FORTIFY_SOURCE -D _FORTIFY_SOURCE=1 -internal-isystem /usr/local/include -internal-isystem /ssd/ednikru/ctu/build_ctu_master/bin/../lib/clang/4.0.0/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -O2 -fdebug-compilation-dir /ssd/ednikru/ctu/tests/vim/src -ferror-limit 19 -fmessage-length 0 -pthread -fobjc-runtime=gcc -fdiagnostics-show-option -vectorize-loops -vectorize-slp -o /ssd/ednikru/ctu/tests/vim/.xtu-out/report-JySHuo.plist -x c /ssd/ednikru/ctu/tests/vim/src/buffer.c
analyze: INFO: 1.	<eof> parser at end of file
analyze: INFO: 2.	While analyzing stack:
analyze: INFO: 	#0 int buf_contents_changed(buf_T *buf)
analyze: INFO: 3.	/ssd/ednikru/ctu/tests/vim/src/buffer.c:6080:9: Error evaluating statement
analyze: INFO: 4.	/ssd/ednikru/ctu/tests/vim/src/buffer.c:6080:9: Error evaluating statement

--- a/lib/AST/ASTImporter.cpp
+++ b/lib/AST/ASTImporter.cpp
@@ -2875,7 +2875,7 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
   DeclarationName SearchName = Name;
   if (!SearchName && D->getTypedefNameForAnonDecl()) {
     SearchName = Importer.Import(D->getTypedefNameForAnonDecl()->getDeclName());
-    IDNS = Decl::IDNS_Ordinary;
+    IDNS |= Decl::IDNS_Ordinary;
   } else if (Importer.getToContext().getLangOpts().CPlusPlus)
     IDNS |= Decl::IDNS_Ordinary;
 
@@ -3000,6 +3000,7 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
     D2->setQualifierInfo(Importer.Import(D->getQualifierLoc()));
     D2->setLexicalDeclContext(LexicalDC);
     LexicalDC->addDeclInternal(D2);
+    D2->setImplicit(D->isImplicit());
     if (D->isAnonymousStructOrUnion())
       D2->setAnonymousStructOrUnion(true);
   }
@@ -3447,6 +3448,7 @@ Decl *ASTNodeImporter::VisitIndirectFieldDecl(IndirectFieldDecl *D) {
 
   ToIndirectField->setAccess(D->getAccess());
   ToIndirectField->setLexicalDeclContext(LexicalDC);
+  ToIndirectField->setImplicit(D->isImplicit());
   Importer.Imported(D, ToIndirectField);
   LexicalDC->addDeclInternal(ToIndirectField);
   return ToIndirectField;

--- a/test/Analysis/Inputs/externalFnMap2.txt
+++ b/test/Analysis/Inputs/externalFnMap2.txt
@@ -1,0 +1,1 @@
+_Z1fi@x86_64 xtu-other.c.ast

--- a/test/Analysis/Inputs/xtu-other.c
+++ b/test/Analysis/Inputs/xtu-other.c
@@ -1,0 +1,14 @@
+
+typedef struct {
+  int a;
+  int b;
+} foobar;
+
+foobar fb;
+
+int f(int i) {
+  if (fb.a) {
+    fb.b = i;
+  }
+  return 1;
+}

--- a/test/Analysis/xtu-main.c
+++ b/test/Analysis/xtu-main.c
@@ -1,0 +1,19 @@
+// RUN: mkdir -p %T/xtudir2
+// RUN: %clang_cc1 -triple x86_64-pc-linux-gnu -emit-pch -o %T/xtudir2/xtu-other.c.ast %S/Inputs/xtu-other.c
+// RUN: cp %S/Inputs/externalFnMap2.txt %T/xtudir2/externalFnMap.txt
+// RUN: %clang_cc1 -triple x86_64-pc-linux-gnu -fsyntax-only -analyze -analyzer-checker=core,debug.ExprInspection -analyzer-config xtu-dir=%T/xtudir2 -analyzer-config reanalyze-xtu-visited=true -verify %s
+
+void clang_analyzer_eval(int);
+
+typedef struct {
+  int a;
+  int b;
+} foobar;
+
+int f(int);
+extern foobar fb;
+
+int main() {
+  clang_analyzer_eval(f(5) == 1); // expected-warning{{TRUE}}
+  return 0;
+}

